rtengine/LUT.h

```diff
@@ -293,72 +298,37 @@ class LUT :
     }
 
 #if defined( __SSE2__ ) && defined( __x86_64__ )
-/*
-    vfloat operator[](vfloat indexv ) const
+    vfloat operator[](vfloat indexv) const
     {
-//      printf("don't use this operator. It's not ready for production");
-        return _mm_setzero_ps();
-
-        // convert floats to ints
-        vint idxv =  _mm_cvttps_epi32( indexv );
-        vfloat tempv, resultv, p1v, p2v;
-        vmask maxmask = vmaskf_gt(indexv, maxsv);
-        idxv = _mm_castps_si128(vself(maxmask, maxsv, _mm_castsi128_ps(idxv)));
-        vmask minmask = vmaskf_lt(indexv, _mm_setzero_ps());
-        idxv = _mm_castps_si128(vself(minmask, _mm_setzero_ps(), _mm_castsi128_ps(idxv)));
-        // access the LUT 4 times and shuffle the values into p1v and p2v
-
-        int idx;
-
-        // get 4th value
-        idx = _mm_cvtsi128_si32 (_mm_shuffle_epi32(idxv, _MM_SHUFFLE(3, 3, 3, 3)));
-        tempv = LVFU(data[idx]);
-        p1v = _mm_shuffle_ps(tempv, tempv, _MM_SHUFFLE(0, 0, 0, 0));
-        p2v = _mm_shuffle_ps(tempv, tempv, _MM_SHUFFLE(1, 1, 1, 1));
-        // now p1v is 3 3 3 3
-        //     p2v is 3 3 3 3
-
-        // get 3rd value
-        idx = _mm_cvtsi128_si32 (_mm_shuffle_epi32(idxv, _MM_SHUFFLE(2, 2, 2, 2)));
-        tempv = LVFU(data[idx]);
-        p1v = _mm_move_ss( p1v, tempv);
-        tempv = _mm_shuffle_ps(tempv, tempv, _MM_SHUFFLE(1, 1, 1, 1));
-        p2v = _mm_move_ss( p2v, tempv);
-        // now p1v is 3 3 3 2
-        //     p2v is 3 3 3 2
+        static_assert(std::is_same<T, float>::value, "This method only works for float LUTs");
+
+        // Clamp and convert to integer values. Extract out of SSE register because all
+        // lookup operations use regular addresses.
+        vfloat clampedIndexes = _mm_max_ps(
+                _mm_setzero_ps(),
+                _mm_min_ps(_mm_set1_ps(maxIndexFloat), indexv));
+        vint indexes = _mm_cvttps_epi32(clampedIndexes);
+        int indexArray[4];
+        _mm_storeu_si128(reinterpret_cast<__m128i*>(&indexArray[0]), indexes);
+
+        // Load data from the table. This reads more than necessary, but there don't seem
+        // to exist more granular operations (though we could try non-SSE).
+        // Cast to int for convenience in the next operation (partial transpose).
+        vint values[4];
+        for (int i = 0; i < 4; ++i) {
+            values[i] = _mm_castps_si128(LVFU(data[indexArray[i]]));
```

Comment:
This potentially reads out of bound. Maybe better to increase the size of the new[] in line 128 by 3?